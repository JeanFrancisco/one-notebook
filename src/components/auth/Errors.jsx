const Errors = ({ errors }) => (
    ( errors !== undefined ) && <>
        <label className="auth__errors">
            { errors }
        </label>
    </>
);

export default Errors;