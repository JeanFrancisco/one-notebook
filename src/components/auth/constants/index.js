export const AUTH_LOGIN = 'AUTH_LOGIN';
export const AUTH_LOGOUT = 'AUTH_LOGOUT';
export const AUTH_SET_ERROR = 'AUTH_ERROR';
export const AUTH_REMOVE_ERROR = 'AUTH_REMOVE_ERROR';

export default {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR,
};