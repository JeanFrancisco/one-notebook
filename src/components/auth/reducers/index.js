import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_REMOVE_ERROR,
    AUTH_SET_ERROR
} from "../constants";

const initialState = {};

export const authReducer = (state = initialState, action) => {

    switch(action.type) {
        case AUTH_LOGIN:
            return {
                uid: action.payload.uid,
                name: action.payload.displayName,
            };

        case AUTH_LOGOUT:
            return {};

        case AUTH_REMOVE_ERROR:
            return {
                ...state,
                errors: undefined,
            };

        case AUTH_SET_ERROR:
            return {
                ...state,
                errors: action.payload,
            };
        default: return state;
    }
}