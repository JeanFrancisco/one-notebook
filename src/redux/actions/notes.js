import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, getDocs, orderBy, updateDoc } from '@firebase/firestore';
import {
    NOTES_NEW_NOTE,
    NOTES_SET_ACTIVE,
    NOTES_DISPLAY_IN_SIDEBAR,
    NOTES_UPDATED_NOTE,
    NOTES_DROP_NOTE,
    NOTES_CLEANING_WHEN_LOGOUT,
    NOTES_CHANGE_ORDER,
} from '../constants/typesNotes';
import { uploadFile } from '../../helpers/files';
import { db as firestoreDB } from '../../config/firebase';

export const actionAddNewNote = () => {
    return async (dispatch, getState) => {
        const { uid: user_uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            urlImage: '',
            createdAt: new Date().getTime(),
        };

        const doc = await addDoc( collection(firestoreDB, `${user_uid}/journal/notes`), newNote);

        const createdNote = {
            id: doc.id,
            ...newNote,
        };

        dispatch( actionNewNote( createdNote ) );
        dispatch( actionSetActiveNote( createdNote ) );
    }
};

export const actionNewNote = ( createdNote ) => ({
    type: NOTES_NEW_NOTE,
    payload: createdNote,
});

export const actionSetActiveNote = ( activeNote ) => ({
    type: NOTES_SET_ACTIVE,
    payload: activeNote,
});

export const actionSavingChanges = () => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;
        const activeNote = getState().notes.active;

        const { id: note_id, ...noteForFirestore } = activeNote;
        await updateDoc( doc(firestoreDB, `${user_uid}/journal/notes/${ note_id }`), noteForFirestore )
            .then(() => {
                dispatch( refreshNote(activeNote) );

                Swal.fire({
                    title: 'Guardada exitosamente',
                    html: '<p><small>Su nota se guardó correctamente</small></p>',
                    icon: 'success',
                    timer: 1300,
                    showConfirmButton: true,
                    confirmButtonText: 'Aceptar',
                    confirmButtonColor: 'limegreen',
                });
            })
            .catch(reason => {
                console.log( reason.code );
            });
    }
}

export const refreshNote = ( updatedNote ) => ({
    type: NOTES_UPDATED_NOTE,
    payload: { ...updatedNote },
});

export const getFirebaseNotes = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const snapNotes = await getDocs( collection(firestoreDB, `${uid}/journal/notes`) );
        const notes = [];

        snapNotes.forEach( snapHijo => {
            notes.push({
                id: snapHijo.id,
                ...snapHijo.data(),
            })
        });

        dispatch( displayNotes(notes) );
    }
}

export const displayNotes = ( notes ) => ({
    type: NOTES_DISPLAY_IN_SIDEBAR,
    payload: notes,
});

export const actionSetAscendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'asc',
    }
});

export const actionSetDescendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'desc',
    }
});

export const actionReorderByRule = ( orderBy ) => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        orderby: orderBy,
    }
});

export const actionStartUploadFile = (file) => {
    return async (dispatch, getState) => {
 
        Swal.fire({
            title: 'Subiendo archivos...',
            text: 'Por favor, espere un momento...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const activeNote = getState().notes.active;
        const { secure_url } = await uploadFile(file);

        dispatch( actionSetActiveNote({ ...activeNote, urlImage: secure_url }) );
        await dispatch( actionSavingChanges() );

        Swal.close();
    }
}

export const actionStartDeletingNote = (note_id) => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;

        try {
            const resp = await deleteDoc( doc( firestoreDB, `${user_uid}/journal/notes/${note_id}`) );

            dispatch( deleteNote( note_id ) );

            Swal.fire({
                icon: 'success',
                timer: 1250,
                title: 'Eliminado',
                text: 'Su nota fue eliminada',
                confirmButtonText: 'Aceptar',
                confirmButtonColor: 'green',
                showCloseButton: true,
            });

        } catch (e) {
            Swal.fire({
                icon: 'error',
                html: 'Error en la conexión con Firestore DB',
                title: 'Error. <small>No pudimos terminar la eliminación</small>',
                footer: `<small>For more details: ${e.code}</small>`,
                confirmButtonText: 'Entendido',
                confirmButtonColor: 'oragered', 
                showCloseButton: true,
            });

            console.error(e);
        }
    }
}

export const deleteNote = (noteID) => ({
    type: NOTES_DROP_NOTE,
    payload: noteID
});

export const cleanupNotesLogout = () => ({
    type: NOTES_CLEANING_WHEN_LOGOUT
});