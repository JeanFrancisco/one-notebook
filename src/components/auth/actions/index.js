import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    GoogleAuthProvider,
    signInWithPopup,
    updateProfile,
    getAuth,
    signOut,
} from "firebase/auth";
import Swal from 'sweetalert2';
import { app } from "../../../config/firebase";
import { cleanupNotesLogout } from "../../../redux/actions/notes";
import { actionFinishLoading, actionStartLoading } from "../../../redux/actions/ui";
import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR
} from "../constants";

export const authLogin = (uid, displayName) => ({
    type: AUTH_LOGIN,
    payload: {
        uid,
        displayName,
    }
});

export const authSignInWithGoogle = () => {
    return async (dispatch) => {

        dispatch( actionStartLoading() );

        const auth = getAuth( app );

        await signInWithPopup( auth, new GoogleAuthProvider() )
        .then( userCredentials => {
            // const { user: { displayName, uid } } = userCredentials;
            
            // dispatch( authLogin(uid, displayName) );
        })
        .catch( reason => {
            Swal.fire({
                icon: 'warning',
                html: '<small>No se pudo iniciar la sesión. <br/> <br/> ¿Alcanzó a ver si ocurrió algo: con el Pop-up de autenticación de Google?</small>',
                title: 'Oups! Lo sentimos.',
                iconHtml: '<i class="fa fa-warning"></i>',
                footer: `<small>Reason Code: ${reason.code}</small>`,
                showCloseButton: true,
                confirmButtonText: 'Aceptar',
            });
        })
        .finally(() => {
            dispatch( actionFinishLoading() );
        });
    };
}

export const authEmailPasswordSignIn = ({ email, password }) => {
    return async dispatch => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await signInWithEmailAndPassword(auth, email, password)
            .then( ({ user }) => {
                // const { uid, displayName } = user;

                // dispatch( authLogin(uid, displayName) );
            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Algo ocurrió mientras iniciabamos tu sesión.</small>',
                    html: `<p><b><small>¿Qué acciones puedes hacer?</small></b></p>
                        <ul style="text-align: left">
                            <li><small>Puedes revisar por favor tu email y contraseña.</small></li>
                            <li><small>También que te puedes conectar a Internet. En especial a los servicios de Google.</small></li>
                        </ul>`,
                    confirmButtonText: 'Entendido',
                    icon: 'warning',
                    iconColor: 'orange',
                    showCloseButton: true,
                    confirmButtonColor: 'orangered',
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                })
            }).finally(() => {
                dispatch( actionFinishLoading() );
            });
    }
}

export const authEmailPasswordSignUp = ({ email, password, username }) => {
    return async (dispatch) => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await createUserWithEmailAndPassword(auth, email, password)
            .then( async userCredentials => {

                const { user } = userCredentials;
                await updateProfile(user, { displayName: username });

                // dispatch( authLogin(user.uid, user.displayName) );

            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Ahora mismo no podemos registrarte.</small>',
                    html: '<p>Por favor... <br/> <small>Regresa más tarde y vuelve a intentarlo ;) .</small><p>',
                    icon: 'info',
                    iconColor: 'orange',
                    confirmButtonText: 'Entiendo',
                    confirmButtonColor: 'limegreen',
                    showCloseButton: true,
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                });
            }).finally(() => {
               dispatch( actionFinishLoading() );
            });
    }
}

export const authFirebaseLogout = () => {
    return async dispatch => {
        const auth = getAuth(app);

        await signOut(auth);

        dispatch( authLogout() );

        dispatch( cleanupNotesLogout() );
    }
}

export const authLogout = () => ({ type: AUTH_LOGOUT });

export const setAuthError = ( msg ) => ({
    type: AUTH_SET_ERROR,
    payload: msg
});

export const removeAuthError = () => ({ type: AUTH_REMOVE_ERROR });