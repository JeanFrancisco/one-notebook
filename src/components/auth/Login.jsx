import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import Errors from './Errors';
import { useForm } from '../../hooks/useForm';
import { authSignInWithGoogle, authEmailPasswordSignIn, removeAuthError, setAuthError } from './actions';

const Login = () => {
    const dispatch = useDispatch();
    const { errors } = useSelector( store => store.auth );
    const { loading } = useSelector( store => store.ui );

    // useEffect(() => {
    //     dispatch( removeAuthError() );
    // }, [ dispatch ]);

    const isFormValid = () => {
        if( validator.isEmpty( formValues.email, ) || validator.isEmpty( formValues.password ) ) {
            dispatch( setAuthError('Por favor, asegúrese de ingresar todos los campos.'))
            return false;
        }

        if( ! validator.isEmail( formValues.email ) ) {
            dispatch( setAuthError('Por favor asegúrese que su email es correcto, el valor ingresado no es válido.') );
            return false;
        }

        if( ! validator.isLength( formValues.password, { min: 6 } )) {
            dispatch( setAuthError('La contraseña es incorrecta, como mínimo debe tener una longitud de 6 caracteres.') );
            return false;
        }

        if( errors !== undefined )
            dispatch( removeAuthError() );

        return true;
    }

    const onSubmit = () => {
        if( isFormValid() ) {
            const action = authEmailPasswordSignIn( formValues );
            dispatch( action );
        }
    }

    const [ formValues, handleInputChanges, handleFormSubmit ] = useForm({
        email: '',
        password: ''
    }, onSubmit)

    const handleSignInWithGoogle = e => {
        const asyncAction = authSignInWithGoogle();
        dispatch( asyncAction );
    }

    return <>
        <h4 className="auth__title">Iniciar Sesión</h4>

        <Errors errors={ errors } />

        <form onSubmit={ handleFormSubmit }>
            <input
                type="email"
                name="email"
                className="auth__input"
                autoComplete="off"
                value={ formValues.email }
                onChange={ handleInputChanges }
                placeholder="Ingresa aquí tu e-mail."/>

            <input
                type="password"
                name="password"
                className="auth__input"
                value={ formValues.password }
                onChange={ handleInputChanges }
                placeholder="Escribe aquí tu contraseña."
                />

            <button
                className="btn btn-primary btn-block auth__button"
                disabled={ loading }
                type="submit">
                    Ingresar
            </button>

            <div className="auth__social-networks">
                <p>Iniciar sesión con redes sociales:</p>

                <div
                    onClick={ handleSignInWithGoogle }
                    disabled={ loading }
                    className="auth__google-btn">
                    <div className="auth__google-icon-wrapper">
                        <img
                            className="auth__google-icon"
                            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                            alt="[google logo]" />
                    </div>
                    <div className="auth__google-btn-text">
                        <b>Sign in with google</b>
                    </div>
                </div>
            </div>
        </form>

        <Link
            className="link"
            to="/auth/register">
            Crear una cuenta nueva.
        </Link>
    </>;
}

export default Login;