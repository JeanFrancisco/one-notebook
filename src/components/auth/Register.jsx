import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import Errors from './Errors';
import { useForm } from '../../hooks/useForm';
import { authEmailPasswordSignUp, removeAuthError, setAuthError } from './actions';
import { useEffect } from 'react';

const Register = () => {

    const dispatch = useDispatch();
    const { errors } = useSelector( state => state.auth );

    useEffect(() => {
        dispatch( removeAuthError() );
    }, [ dispatch ]);

    const isFormValid = () => {

        if( validator.isEmpty(formValues.username) ) {
            dispatch( setAuthError('Su nombre de usuario no puede ser un valor vacío') );
            return false;
        }

        if( ! validator.isEmail(formValues.email) ) {
            dispatch( setAuthError('El email no es válido') );
            return false;
        }

        if( ! validator.isLength(formValues.password, { min: 6 } ) || formValues.password !== formValues.confirmpassword ) {
            dispatch( setAuthError(`Revise su contraseña. Recuerde que no puede tener menos de 6 carácteres y
            debe coincidir con la confirmación de password`) );
            return false;
        }

        if( errors !== undefined )
            dispatch( removeAuthError() );

        return true;
    }

    const onSubmit = () => {
        if( isFormValid() )
            dispatch( authEmailPasswordSignUp(formValues) );
    };

    const [ formValues, handleInputChange, handleSubmit ] = useForm({
        username: 'Francisco',
        email: 'username@example.org',
        password: '122345',
        confirmpassword: '398174',
    }, onSubmit );

    return <>
        <h4 className="auth__title">Registrar</h4>

        <Errors errors={ errors } />

        <form onSubmit={ handleSubmit }>
            <input
                type="text"
                name="username"
                className="auth__input"
                autoComplete="off"
                placeholder="Enter your username"
                onChange={ handleInputChange }
                value={ formValues.username }
                />

            <input
                type="email"
                name="email"
                className="auth__input"
                autoComplete="off"
                placeholder="Enter your e-mail"
                onChange={ handleInputChange }
                value={ formValues.email }
                />

            <input
                type="password"
                name="password"
                className="auth__input"
                placeholder="Type your password"
                onChange={ handleInputChange }
                value={ formValues.password }
                />

            <input
                type="password"
                name="confirmpassword"
                className="auth__input"
                placeholder="Please confirm your password"
                onChange={ handleInputChange }
                value={ formValues.confirmpassword }
                />

            <button
                className="btn btn-primary btn-block auth__button"
                type="submit">
                    Registrarme
            </button>
        </form>

        <Link
            className="link"
            to="/auth/login">
            ¿Ya estás registrado? Inicia Sesión.
        </Link>
    </>
}

export default Register;