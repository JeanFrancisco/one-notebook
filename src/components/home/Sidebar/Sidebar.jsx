import { useDispatch, useSelector } from "react-redux";
import JournalList from "../JournalList/JournalList";
import { authFirebaseLogout } from "../../auth/actions";
import { actionAddNewNote } from "../../../redux/actions/notes";

const Sidebar = () => {
    const dispatch = useDispatch();
    const user = useSelector( store => store.auth );

    const handleLogout = () => {
        const asyncAction = authFirebaseLogout();
        dispatch( asyncAction );
    }

    const handleClickNewEntry = () => {
        const asyncAction = actionAddNewNote();
        dispatch( asyncAction );
    }

    return <aside className="sidebar">
        <div className="sidebar__user-info">
            <i className="sidebar__user-icon far fa-moon"></i>
            <p className="sidebar__user-name">
                { user.name }
            </p>

            <button
                className="btn sidebar__user-loggout"
                onClick={ handleLogout } >
                Logout
            </button>
        </div>

        <div className="sidebar__new-entry">
            <button
                onClick={ handleClickNewEntry }
                className="btn">
                <div className="sidebar__new-entry-icon fa-5x">
                    <i className="far fa-calendar"></i>
                    <i className="fas fa-plus" data-fa-transform="shrink-8 left-9 down-5"></i>
                </div>
                <span className="sidebar__new-entry-txt">
                    New Entry
                </span>
            </button>
        </div>

        <JournalList />
    </aside>
}

export default Sidebar;