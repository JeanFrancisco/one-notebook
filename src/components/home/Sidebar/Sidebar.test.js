import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { actionAddNewNote } from "../../../redux/actions/notes";
import { authFirebaseLogout } from "../../auth/actions";
import Sidebar from "./Sidebar";

jest.mock('../../auth/actions', () => ({
    authFirebaseLogout: jest.fn(),
}));

jest.mock('../../../redux/actions/notes', () => ({
    actionAddNewNote: jest.fn(),
}));

const initialState = {
    auth: {
        uid: 'TEST_UNKNOWN_UID',
        name: 'Jane Doe',
    },
    notes: {
        notes: [],
        active: null,
    }
};

const mockStore = configureStore([ thunk ]);

const store = mockStore(initialState);

store.dispatch = jest.fn();

describe('Pruebas al componente <Sidebar />.jsx', () => {
    const wrapper = mount(<Provider store={ store }>
        <Sidebar />
    </Provider>);

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('Debe de renderizar el componente haciendo match con el snapshot por defecto.', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe llamar a la acción de authFirebaseLogout()', () => {
        wrapper.find('.sidebar__user-loggout').prop('onClick')();

        expect( authFirebaseLogout ).toHaveBeenCalledTimes(1);
    });

    test('Debe llamar a la accion actionAddNewNote() cuando da click al botón agregar New Entry', () => {
        wrapper.find('.sidebar__new-entry>.btn').prop('onClick')();

        expect( actionAddNewNote ).toHaveBeenCalledTimes(1);
    });
});