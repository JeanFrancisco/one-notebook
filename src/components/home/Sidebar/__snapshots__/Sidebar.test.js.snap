// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <Sidebar />.jsx Debe de renderizar el componente haciendo match con el snapshot por defecto. 1`] = `
<aside
  className="sidebar"
>
  <div
    className="sidebar__user-info"
  >
    <i
      className="sidebar__user-icon far fa-moon"
    />
    <p
      className="sidebar__user-name"
    >
      Jane Doe
    </p>
    <button
      className="btn sidebar__user-loggout"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="sidebar__new-entry"
  >
    <button
      className="btn"
      onClick={[Function]}
    >
      <div
        className="sidebar__new-entry-icon fa-5x"
      >
        <i
          className="far fa-calendar"
        />
        <i
          className="fas fa-plus"
          data-fa-transform="shrink-8 left-9 down-5"
        />
      </div>
      <span
        className="sidebar__new-entry-txt"
      >
        New Entry
      </span>
    </button>
  </div>
  <div
    className="journal__list"
  >
    <div
      className="journal__list-order-form"
    >
      <label
        className="journal__list-order-title"
      >
         Order by: 
      </label>
      <input
        checked={false}
        id="alphabetic"
        name="orderBy"
        onChange={[Function]}
        type="radio"
        value="alphabetic"
      />
      <label
        htmlFor="alphabetic"
      >
        Alphabetically
      </label>
      <input
        checked={false}
        id="cronologic"
        name="orderBy"
        onChange={[Function]}
        type="radio"
        value="cronologic"
      />
      <label
        htmlFor="cronologic"
      >
        Cronologically
      </label>
      <button
        className="btn"
        onClick={[Function]}
      >
        <i
          className="fa fa-sort-amount-up"
          id="btn-sort-icon"
        />
      </button>
    </div>
  </div>
</aside>
`;
